modules/ml/src/main/java/org/apache/ignite/ml/tree/randomforest/RandomForestTrainer.java

```diff
@@ -17,45 +17,815 @@
+     * @param sampleId Sample id.
+     * @param meta Bucket Meta.
+     * @return impurity computer
+     */
+    protected abstract S createImpurityComputer(int sampleId, BucketMeta meta);
+
+    /**
+     * Merge leaf statistics from several data partitions.
+     *
+     * @param left Left.
+     * @param right Right.
+     * @return merged leaf impurity statistics.
+     */
+    Map<NodeId, NodeStatistics> reduceImpurityStatistics(Map<NodeId, NodeStatistics> left,
+        Map<NodeId, NodeStatistics> right) {
+
+        if (left == null)
+            return right;
+        if (right == null)
+            return left;
+
+        Map<NodeId, NodeStatistics> result = new HashMap<>(left);
+        for (NodeId key : right.keySet()) {
+            NodeStatistics rightVal = right.get(key);
+            if (!result.containsKey(key))
+                result.put(key, rightVal);
+            else
+                result.get(key).addOtherStatistics(rightVal.perFeatureStatistics);
+        }
+
+        return result;
+    }
+
+    /**
+     * Aggregator of normal distribution statistics for continual features.
+     */
+    static class NormalDistributionStats {
+        /** Min value. */
+        private final double min;
+
+        /** Max value. */
+        private final double max;
+
+        /** Sum of value squares. */
+        private final double sumOfSquares;
+
+        /** Sum of values. */
+        private final double sumOfValues;
+
+        /** Count of objects. */
+        private final long n;
+
+        /**
+         * Creates an instance of NormalDistributionStats.
+         *
+         * @param min Min.
+         * @param max Max.
+         * @param sumOfSquares Sum of squares.
+         * @param sumOfValues Sum of values.
+         * @param n N.
+         */
+        public NormalDistributionStats(double min, double max, double sumOfSquares, double sumOfValues, long n) {
+            this.min = min;
+            this.max = max;
+            this.sumOfSquares = sumOfSquares;
+            this.sumOfValues = sumOfValues;
+            this.n = n;
+        }
+
+        /**
+         * Returns sum of normal distribution statistics.
+         *
+         * @param stats Stats.
+         * @return sum of normal distribution statistics.
+         */
+        public NormalDistributionStats plus(NormalDistributionStats stats) {
+            return new NormalDistributionStats(
+                Math.min(this.min, stats.min),
+                Math.max(this.max, stats.max),
+                this.sumOfSquares + stats.sumOfSquares,
+                this.sumOfValues + stats.sumOfValues,
+                this.n + stats.n
+            );
+        }
+
+        /**
+         * Mean value.
+         */
+        public double mean() {
+            return sumOfValues / n;
+        }
+
+        /**
+         * Variance value.
+         */
+        public double variance() {
+            double mean = mean();
+            return (sumOfSquares / n) - mean * mean;
```

Comment:
Are you sure in that? I don't understand how you get it from `1/n * (x_i - mean)^2 for all i`.

Reply:
see https://en.wikipedia.org/wiki/Variance